tests/tf/model/test_model.py

```diff
@@ -85,83 +85,65 @@ def test_to_tf_model_from_config(yoochoose_schema, tf_yoochoose_like, config_cls
+def test_output_shape_mode_eval(tf_yoochoose_like, tf_next_item_prediction_model, masking):
 
-    input_module = tr.TabularSequenceFeatures.from_schema(
-        yoochoose_schema,
-        max_sequence_length=20,
-        continuous_projection=64,
-        d_output=64,
-        masking=masking,
-    )
-
-    transformer_config = tr.XLNetConfig.build(d_model=64, n_head=8, n_layer=2, total_seq_length=20)
-    body = tr.SequentialBlock(
-        [
-            input_module,
-            tr.TransformerBlock(transformer_config, masking=input_module.masking),
-        ]
-    )
-    task = tr.NextItemPredictionTask(weight_tying=True)
-    model = task.to_model(body=body)
+    model = tf_next_item_prediction_model(masking=masking)
 
     out = model(tf_yoochoose_like, training=False)
     assert out.shape[0] == tf_yoochoose_like["item_id/list"].shape[0]
 
 
 @pytest.mark.parametrize("masking", ["causal", "mlm"])
-def test_next_item_fit(tf_yoochoose_like, yoochoose_schema, masking, run_eagerly=True):
+@pytest.mark.parametrize("run_eagerly", [True, False])
+def test_next_item_fit_model(
+    tf_yoochoose_like, tf_next_item_prediction_model, masking, run_eagerly
+):
 
-    input_module = tr.TabularSequenceFeatures.from_schema(
-        yoochoose_schema,
-        max_sequence_length=20,
-        continuous_projection=64,
-        d_output=64,
-        masking=masking,
-    )
-
-    transformer_config = tr.XLNetConfig.build(d_model=64, n_head=8, n_layer=2, total_seq_length=20)
-    body = tr.SequentialBlock(
-        [
-            input_module,
-            tr.TransformerBlock(transformer_config, masking=input_module.masking),
-        ]
-    )
-    task = tr.NextItemPredictionTask(weight_tying=True)
-    model = task.to_model(body=body)
+    model = tf_next_item_prediction_model(masking=masking)
     model.compile(optimizer="adam", run_eagerly=run_eagerly)
 
     dataset = tf.data.Dataset.from_tensor_slices(
         (tf_yoochoose_like, tf_yoochoose_like["item_id/list"])
     ).batch(50)
-
     losses = model.fit(dataset, epochs=5)
     metrics = model.evaluate(tf_yoochoose_like, tf_yoochoose_like["item_id/list"], return_dict=True)
 
-    assert len(metrics.keys()) == 6
+    assert len(metrics.keys()) == 9
     assert len(losses.epoch) == 5
     assert all(loss >= 0 for loss in losses.history["loss"])
+    assert losses.history["loss"][-1] < losses.history["loss"][0]
+
+
+@pytest.mark.parametrize("run_eagerly", [True, False])
+@pytest.mark.parametrize("config", config_classes)
+def test_save_and_load_transformer_model(
+    tf_next_item_prediction_model,
+    tf_yoochoose_like,
+    config,
+    run_eagerly,
+):
+    model = tf_next_item_prediction_model(config=config)
+    model.compile(optimizer="adam", run_eagerly=run_eagerly)
+
+    dataset = tf.data.Dataset.from_tensor_slices(
+        (tf_yoochoose_like, tf_yoochoose_like["item_id/list"])
+    ).batch(50)
+
+    inputs = next(iter(dataset))[0]
+    model._set_inputs(inputs)
```

Comment:
Why are we calling _set_inputs() method here?

Reply:
To save the model, we need to set the input shapes of all sub-modules, and this is automatically done when we call the `fit` or `call` methods of the model. But here, we only focus on the saving part, so we needed to explicitly call `model._set_inputs(...)` to set the shapes.